### PR TITLE
Add merged *nix agent

### PR DESCRIPTION
**Pull request by rawiriblundell**
In PR #21, Sven mentioned the desire to reduce agent counts down to two and to eliminate bashisms.

This PR provides a merged agent script with all but a small handful of non-POSIX-isms eliminated.

It features:

* Added vim standards line at top of script
* Updated homepage URL in copyright header
* Portability fixes like ensuring that commonly called ENV vars are available, and one or two uncommon ones
* Auto-builds PATH, primarily to ensure that Solaris uses XPG4 where possible
* A function to handle the main portability issues of 'echo'
* Addition of MK_VERSION variable.
* Addition of MK_OSSTR variable.  This one is pretty important.
* A fix for timeout/waitmax to map it to the busybox variant of timeout (if found)
* Because 'local' is not POSIX, most functions now have a leading underscore for the variables (e.g. "${_example_var}") to denote local scope, and these are explicitly unset.  Implementing this script-wide is incomplete and ongoing.
* I've made an attempt to provide the desired arg handling in the run_cached() function via getopts
* The openbsd network interface audit code was a portability disaster, so I've completely overhauled it, it tests fine on a couple of openbsd vm's.
* Other large blocks of code were portability disasters requiring similar overhauls^[1]
* I've added the basis of a section_iostat() function to provide those metrics if /proc/diskstats doesn't exist.  This needs more work.  openbsd does not appear to use the same output format as others, so it's ringfenced.
* Almost everything is a function now
* Additional comments on some non-obvious code
* Countless other things

Things that are known-broken and needing attention:

* Exporting functions is not POSIX, these might need to be written to temp files
* ~~Encryption, as process substitution is not defined in POSIX~~
* ~~RTC_SECRET variable is modified within a subshell~~

To-do:

* So many things.  Some are even noted in the code (search for "TO-DO')

/Edit: Finally, I expect that if this accepted and tested sufficiently, that at some point the '.merged' will be dropped and it will simply be 'check_mk_agent'.  This 'extension' is simply intended as a differentiator until that point.

^[1]I expect that this will break a few things, so I recommend testing.  Lots of testing.  It would be insanity to think that 11 years of shell script development could be perfectly untangled at the first attempt, but it's a start.

### CODE REVIEW

**Comment by rawiriblundell** [comment]
One quick comment about the shebang:  /bin/sh is going to fail miserably on Solaris 10 or older, as it's a pre-POSIX shell on those systems.  For everything else, /bin/sh should be some variant of ksh, ash, bash or dash, which are all POSIX compliant or near enough.  And even then, on older versions of supported systems (and we're talking 1990-1994 era... WHY would you have that in production???), they are all going to come packaged with a ksh88 variant.

In short: if you're using this on Solaris, it'll blow up.  Change the shebang to `/usr/xpg4/bin/sh` or `/usr/bin/ksh` (or /path/to/bash if you have it installed) and try again.  On everything else - 
 bugs/errors/regressions aside - it should just work (famous last words?)

**Comment by jonaskluger** [comment]
Hello Rawiri,

thank you for this agent file. We will have a closer look at your patch and discuss internally on how and if we can implement it into the official code base.

Internal Ref: CMK-2913

**Comment by rawiriblundell** [comment]
Hi, the force push a few days ago appears to have broken every existing PR's conflict checks.  Are you guys able to fix this mess, or would it be easier if the PR's were recreated?

**Comment by LarsMichelsen** [comment]
Sorry, this mess was made necessary by a required force-push. We had to make some history adjustments to our Github repository (yes, you don't actually...).

To fix this, you would have to rebase your local clones to the upstream branch and skip all commits except your local commits.

**Comment by rawiriblundell** [comment]
Thanks Mike,
That confirms that the one outstanding portability issue actually is an issue:

```
# Make run_cached available for subshells (plugins, local checks, etc.)
# TO-DO: Check if this export is still required
export -f run_cached
+ export -f run_cached
/usr/bin/check_mk_agent.merged[651]: export: bad option(s)
```

i.e. `export -f` is not portable.

I'll have a think about a solution for this.

**Comment by mike1098** [comment]
@rawiriblundell As mentioned in the other pull request from you. In AIX 7 /bin/sh and /usr/bin/sh is linked to the korn shell.

**Comment by rawiriblundell** [comment]
Hi @mike1098 , Thanks for that confirmation :)

Small backstory:  At a previous role, I was tasked with writing a shell based *nix auditing tool, because it didn't seem there were any tools out there that really met our needs.  So this tool runs on AIX, HPUX, Solaris, FreeBSD and various distributions of Linux.  It logically sits somewhere at the intersection of `scc` and `etckeeper`, and it takes a modular approach rather than the monolithic approach taken by the likes of `cfg2html` and this merged script.  So, basically, there's tens of thousands of lines of shell code for auditing various unices out there, in use, that I've written.

Anyway, [this](https://www.in-ulm.de/~mascheck/various/shells/) was an extremely useful read to convince me that `ksh` is _the_ sane common denominator, rather than pure bourne or `perl` or anything else.  So that's the target that I largely wrote the auditing tool towards, and hence my comment above :)

**Comment by rawiriblundell** [comment]
Closing due to lack of engagement from tribe29.

**Comment by mo-ki** [comment]
I am sorry. And I know that doesn't help the cause, but I still want to say that the lack of engagement results from a lack of time, not a lack of interest (speaking for myself, there).
I've come here frequently for reference, but I think the only way we can even come close to this (not even to mention your fork) is by incremental changes. You may have noticed my rather blunt attempt to at least stop five of the agents from diverging further, and be able to incrementally align them.
Closed or not, your first comment contains a valid to-do list.

**Comment by rawiriblundell** [comment]
@mo-ki I have also attempted the incremental approach and have had mixed, but mostly disappointing results :( 

And, I mean, I was in the middle of creating a PR for an incremental change when tribe29 literally told me to not bother:

https://forum.checkmk.com/t/clarity-required-on-prs-and-bugfixes/23988

Otherwise, keep up the good work (speaking to you directly, there :) )
